src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1070,16 +1116,59 @@ CitusHasBeenLoaded(void)
 			 * in case of extension update.
 			 */
 			installedExtensionVersion = NULL;
+			installedVersionCompatible = NULL;
 		}
 	}
 
-	if (extensionLoaded)
+	return extensionLoaded;
+}
+
+
+/*
+ * ErrorIfDistributedTableLookupOnVersionMismatch checks if given relationId is belong
+ * to a distributed table. If it is and if there is version mismatch, this function
+ * errors out.
+ */
+static void
+ErrorIfDistributedTableLookupOnVersionMismatch(Oid relationId)
+{
+	/*
+	 * We will not use cache if there is a version mismatch, because while
+	 * creating the cache, we assume our metadata tables are in particular
+	 * structure. However, in version mismatch, this is not guaranteed, thus
+	 * we will read from pg_dist_partition table and we will only assume
+	 * specific details which are most likely do not change in long term.
+	 */
+	if (!EnableVersionChecks)
 	{
-		ErrorIfAvailableVersionMismatch();
-		ErrorIfInstalledVersionMismatch();
+		if (!AvailableVersionCompatible())
+		{
+			if (CheckIfDistributedTableFromCatalog(relationId))
+			{
+				ereport(ERROR, (errmsg("loaded Citus library version differs from latest "
+									   "available extension version"),
+								errdetail("Loaded library requires %s, but the latest "
+										  "control file specifies %s.",
+										  CITUS_MAJORVERSION,
+										  AvailableExtensionVersion()),
+								errhint("Restart the database to load the latest Citus "
+										"library.")));
+			}
+		}
+		else if (!InstalledVersionCompatible())
+		{
+			if (CheckIfDistributedTableFromCatalog(relationId))
+			{
+				ereport(ERROR, (errmsg("loaded Citus library version differs from "
+									   "installed extension version"),
+								errdetail("Loaded library requires %s, but the installed "
+										  "extension version is %s.", CITUS_MAJORVERSION,
+										  InstalledExtensionVersion()),
+								errhint("Run ALTER EXTENSION citus UPDATE and try "
+										"again.")));
+			}
+		}
```

Comment:
What I'm saying is that you shouldn't individually cache the results of the different relevant versions, but instead have a single cache: citusVersionsCompatible. If that's false, always perform both checks in an uncached manner.